File: backend/internal/model/vacinacao.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Vacinacao representa um registro de vacinacao de um lote.
type Vacinacao struct {
	ID               uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	LoteID           uuid.UUID `gorm:"type:uuid;not null;index:idx_vacinacoes_lote" json:"lote_id" validate:"required"`
	Data             time.Time `gorm:"type:date;not null" json:"data" validate:"required"`
	Vacina           string    `gorm:"type:varchar(200);not null" json:"vacina" validate:"required,max=200"`
	LoteProduto      *string   `gorm:"type:varchar(100)" json:"lote_produto,omitempty" validate:"omitempty,max=100"`
	ViaAdministracao *string   `gorm:"type:varchar(50)" json:"via_administracao,omitempty" validate:"omitempty,max=50"`
	Responsavel      *string   `gorm:"type:varchar(200)" json:"responsavel,omitempty" validate:"omitempty,max=200"`
	Observacao       *string   `gorm:"type:text" json:"observacao,omitempty"`
	CreatedAt        time.Time `gorm:"not null;default:now()" json:"created_at"`

	// Relacionamentos
	Lote Lote `gorm:"foreignKey:LoteID" json:"-"`
}

func (Vacinacao) TableName() string {
	return "vacinacoes"
}

func (v *Vacinacao) BeforeCreate(tx *gorm.DB) error {
	if v.ID == uuid.Nil {
		v.ID = uuid.New()
	}
	return nil
}

File: flutter_app/lib/features/sanidade/domain/entities/vacinacao.dart
/// Registro de vacinacao aplicada em um lote.
///
/// Matches backend DTO VacinacaoResponse:
/// - vacina (not nome)
/// - data (not data_aplicacao)
/// - lote_produto (not lote)
/// - via_administracao
class Vacinacao {
  final String id;
  final String loteId;
  final String data;
  final String vacina;
  final String? loteProduto;
  final String? viaAdministracao;
  final String? responsavel;
  final String? observacao;
  final String createdAt;

  const Vacinacao({
    required this.id,
    required this.loteId,
    required this.data,
    required this.vacina,
    this.loteProduto,
    this.viaAdministracao,
    this.responsavel,
    this.observacao,
    required this.createdAt,
  });

  factory Vacinacao.fromJson(Map<String, dynamic> json) {
    return Vacinacao(
      id: json['id'] as String,
      loteId: json['lote_id'] as String,
      data: json['data'] as String,
      vacina: json['vacina'] as String,
      loteProduto: json['lote_produto'] as String?,
      viaAdministracao: json['via_administracao'] as String?,
      responsavel: json['responsavel'] as String?,
      observacao: json['observacao'] as String?,
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lote_id': loteId,
      'data': data,
      'vacina': vacina,
      if (loteProduto != null) 'lote_produto': loteProduto,
      if (viaAdministracao != null) 'via_administracao': viaAdministracao,
      if (responsavel != null) 'responsavel': responsavel,
      if (observacao != null) 'observacao': observacao,
      'created_at': createdAt,
    };
  }
}
